frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
-    removeContainerRequest(
-        ResourceDescriptor.fromResource(request.getCapability()),
-        request.getNodes());
-  }
-
-  @VisibleForTesting
-  public synchronized void removeContainerRequest(ResourceDescriptor resource, List<String> nodeHosts) {
-    for (String nodeHost : nodeHosts) {
-      if (candidateNodes.containsKey(nodeHost)) {
-        candidateNodes.get(nodeHost).removeContainerRequest(resource);
-      } else {
-        LOGGER.logWarning("removeContainerRequest: Node is no longer a candidate: %s", nodeHost);
+  private void filterNodesByNodeLabel(String requestNodeLabel) {
+    if (requestNodeLabel != null) {
+      for (int i = filteredNodes.size(); i > 0; i--) {
+        Set<String> availableNodeLabels = allNodes.get(filteredNodes.get(i - 1)).getLabels();
+        if (!HadoopUtils.matchNodeLabel(requestNodeLabel, availableNodeLabels)) {
+          LOGGER.logDebug("NodeLabel does not match: Node: [%s] Request NodeLabel: [%s]",
+              filteredNodes.get(i - 1), requestNodeLabel);
+          filteredNodes.remove(i - 1);
+        }
       }
     }
   }
 
-  public synchronized SelectionResult select(
-      ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType)
-      throws NotAvailableException {
-    LOGGER.logInfo(
-        "select: Given Request: Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s]",
-        requestResource, requestNodeLabel, requestNodeGpuType);
-
-    // ClusterConfiguration is ready when this method is called, i.e. it is not null here.
-    ClusterConfiguration clusterConfiguration = am.getClusterConfiguration();
-    Map<String, NodeConfiguration> configuredNodes = clusterConfiguration.getNodes();
-
-    // Start to select from candidateNodes
-    SelectionResult selectionResult = null;
-    for (Node node : candidateNodes.values()) {
-      String nodeHost = node.getHost();
-      String logPrefix = String.format("select: [%s]: Test Node: ", nodeHost);
-      String rejectedLogPrefix = logPrefix + "Rejected: Reason: ";
-
-      LOGGER.logDebug(logPrefix + "Start: %s", node);
-
-      // Test NodeLabel
-      Set<String> availableNodeLabels = node.getLabels();
-      if (!HadoopUtils.matchNodeLabel(requestNodeLabel, availableNodeLabels)) {
-        LOGGER.logDebug(rejectedLogPrefix +
-                "NodeLabel does not match: Request NodeLabel: [%s], Available NodeLabel: [%s]",
-            requestNodeLabel, CommonExts.toString(availableNodeLabels));
-        continue;
-      }
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        if (configuredNodes != null) {
+  private void filterNodesByGpuType(String requestNodeGpuType) {
+    if (requestNodeGpuType != null) {
+      Map<String, NodeConfiguration> configuredNodes = am.getClusterConfiguration().getNodes();
+      if (configuredNodes != null) {
+        for (int i = filteredNodes.size(); i > 0; i--) {
```

Comment:
Miss else:
          LOGGER.logWarning(logPrefix +
                  "Configured Nodes is not found in ClusterConfiguration: Ignore Request NodeGpuType: [%s]",
              requestNodeGpuType);

Reply:
"Configured Nodes is not found in ClusterConfiguration: Ignore Request NodeGpuType: [%s]",

---
In reply to: [174698174](https://github.com/Microsoft/pai/pull/318#discussion_r174698174) [](ancestors = 174698174)

Reply:
This is not ConfiguredNodes, it is filtered Node.

---
In reply to: [175412947](https://github.com/Microsoft/pai/pull/318#discussion_r175412947) [](ancestors = 175412947,174698174)